src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -11469,119 +11591,373 @@ def update_tag(
+                                )
+                            )
+                    elif isinstance(resource, RunTemplateSchema):
+                        scope_id = None
+                        if resource.source_deployment:
+                            if resource.source_deployment.pipeline_id:
+                                scope_id = (
+                                    resource.source_deployment.pipeline_id
+                                )
+                                scope_ids[
+                                    TaggableResourceTypes.RUN_TEMPLATE
+                                ].append(scope_id)
+
+                        older_templates = self.list_run_templates(
+                            RunTemplateFilter(
+                                id=f"notequals:{resource.id}",
+                                pipeline_id=scope_id,
+                                tags=[tag_schema.name],
+                            )
+                        )
+                        if older_templates.items:
+                            detach_resources.append(
+                                TagResourceRequest(
+                                    tag_id=tag_schema.id,
+                                    resource_id=older_templates.items[0].id,
+                                    resource_type=TaggableResourceTypes.RUN_TEMPLATE,
+                                )
+                            )
+                    else:
+                        logger.debug(
+                            "Exclusive tag functionality only works for "
+                            "templates, for pipeline runs (within the scope of "
+                            "pipelines) and for artifact versions (within the "
+                            "scope of artifacts)."
+                        )
+
+                    # Check for duplicate IDs in any of the scope_ids list
+                    for resource_type, id_list in scope_ids.items():
+                        if len(id_list) != len(set(id_list)):
+                            raise ValueError(
+                                f"You are trying to attach an exclusive tag to "
+                                f"multiple {resource_type.value}s within the "
+                                "same scope. This is not allowed."
+                            )
+
+                    if detach_resources:
+                        self._delete_tag_resource_schemas(
+                            tag_resources=detach_resources,
+                            session=session,
+                            # Don't commit the session here, because we want
+                            # to have one big mega-transaction.
+                            commit=False,
+                        )
+
+                tag_resources_to_create.add(
+                    (tag_schema.id, resource.id, resource_type)
+                )
+            try:
+                session.commit()
+            except IntegrityError:
```

Comment:
When would this commit actually fail with an integrity error? The properties of the `exclusive` attribute of a tag are not enforced in the DB unless I missed it?

Reply:
@schustmi this would fail with an integrity error if we have overlapping tag resources - i.e. in case of race-conditions. When this happens, we rollback and try again. The next loop will detect and reuse the existing entries instead of trying to create them. 

This is the equivalent of get-or-create, but with multiple entries instead of just one.

Reply:
Makes sense. I would have thought doing it 1 by 1 with a separate commit would be the much simpler solution to both understand and also safer (= no retries needed). Sure it's more commits but I don't think this will get called with huge amounts of entities either at once, so not sure that would've been a problem.

The more pressing issue is that this is simply no solution to the exclusivity feature of tags at all when it comes to race conditions.

Reply:
@schustmi separate commits means that a lot of things will be left over if the operation as a whole fails. The advantage of one mega-transaction is that it's an all-or-none result.

Reply:
@stefannica At least for all the addition of tag resources, there is not such thing as a leftover I think? We want them all tagged, if someone else tags them before, that's okay as well.

For the removal: We want all the tags removed. If someone else removed them for us, that's okay as well.
In the end it all comes down to the treatment of the `exclusive` property.